jdi-light/src/main/java/com/epam/jdi/light/driver/WebDriverFactory.java

```diff
@@ -32,91 +30,138 @@
     private static void setRunDrivers(MapArray<String, WebDriver> map) {
-        if (SINGLE_THREAD) {
-            RUN_DRIVERS = map;
-        } else {
+        if (MULTI_THREAD) {
             THREAD_RUN_DRIVERS.set(map);
+        } else {
+            RUN_DRIVERS.clear();
+            RUN_DRIVERS.addAll(map);
         }
     }
-    public static WebDriver getDriverByName(String driverName) {
-        logger.debug("getDriverByName(%s)", driverName);
-        Lock lock = new ReentrantLock();
-        if (!SWITCH_THREAD && INIT_DRIVER != null && INIT_THREAD_ID != currentThread().getId()) {
-            logger.debug("SWITCH_THREAD = true");
-            setRunDrivers(map($(driverName, INIT_DRIVER)));
-            SWITCH_THREAD = true;
-            return INIT_DRIVER;
+    private static WebDriver registerNewDriver(String driverName, WebDriver driver, MapArray<String, WebDriver> drivers) {
+        logger.trace("registerNewDriver >> " + driver);
+        drivers.add(driverName, driver);
+        return driver;
+    }
+    private static WebDriver registerNewDriver(String driverName, MapArray<String, WebDriver> drivers) {
+        Lock locker = new ReentrantLock();
+        locker.lock();
+        try {
+            if (!DRIVERS.has(driverName) && DRIVER.types.has(driverName))
+                useDriver(driverName, () -> DRIVER.types.get(driverName).getDriver());
+            if (!DRIVERS.has(driverName))
+                throw exception("Can't get driver '%s'. Please use drivers from JDISettings.DRIVER.types list. " +
+                        "Or add your own driver with WebDriverFactory.useDriver(name,() -> WebDriver) method.");
+            WebDriver driver = getValidDriver(driverName);
+            driver = DRIVER.setup.execute(driver);
+            registerNewDriver(driverName, driver, drivers);
+            return driver;
+        } catch (Exception ex) {
+            throw exception(ex, "Can't run and register new driver '%s'", driverName);
+        } finally {
+            locker.unlock();
         }
-        if (!DRIVERS.has(driverName)) {
-            logger.debug("Has no driver");
-            useDriver(driverName);
+    }
+    private static WebDriver getValidDriver(String driverName) {
+        boolean goodDriver = false;
+        WebDriver driver = null;
+        Timer timer = new Timer(10);
+        while (!goodDriver && !timer.timeoutPassed()) {
+            try {
+                if (driver != null) {
+                    driver.quit();
+                }
+                logger.trace("getValidDriver: Getting driver...");
+                driver = DRIVERS.get(driverName).execute();
+                goodDriver = isNotBlank(driver.getCurrentUrl());
+                logger.trace("getValidDriver: Get driver success");
+            } catch (Throwable ignore) {
+                logger.trace("getValidDriver: Get driver failed");
+            }
         }
+        return driver;
+    }
+    public static boolean GETTING_DRIVER = false;
```

Comment:
Codacy found an issue: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6397461)